File: 2015/Python/14.py
from parse import parse

def generator(input) : 
    return [list(parse("{} can fly {:d} km/s for {:d} seconds, but then must rest for {:d} seconds.", r))[1:] for r in input.splitlines()]

def part_1(input) : 
    return solver(input, new_rule = False) 

def part_2(input) : 
    return solver(input, new_rule = True)
    
def solver(input, new_rule) :
    bonus = [0 for _ in range(len(input))]
    for s in range(2503-2502 * new_rule, 2504):
        ranking = []
        for speed, duration, rest in input : 
            q = s // (duration + rest)
            r = min(duration, s % (duration + rest))
            ranking.append(speed * (r + q * duration))
        bonus = [sum(x)  for x in zip(bonus, [item==max(ranking) for item in ranking])]
    return [max(ranking) , max(bonus)][new_rule]

File: 2015/Rust/src/bin/14.rs
aoc_2015::main!();

fn generator(input : &str) -> Vec<(usize, usize, usize)> {
    input   
        .lines()
        .map(|l| l.split_whitespace().collect_vec())
        .map(|v| (v[3].parse::<usize>().unwrap(), v[6].parse::<usize>().unwrap(), v[13].parse::<usize>().unwrap()))
        .collect_vec()
}

fn part_1(input : Vec<(usize, usize, usize)>) -> usize {
    *solver(input.clone(), 2503)
        .iter()
        .max()
        .unwrap()
}

fn part_2(input : Vec<(usize, usize, usize)>) -> usize {
    let mut bonus = [0 ; 9] ; 
    (1..2503)
        .for_each(|s| {
            let score = solver(input.clone(),s);
            score.iter()
                .positions(|v| v == score.iter().max().unwrap())
                .for_each(|v|  bonus[v] += 1) ;
        });
        *bonus.iter().max().unwrap()
    }

fn solver(input : Vec<(usize, usize, usize)>, seconds : usize) -> Vec<usize> {
    input
    .iter()
    .map(|(speed, time, rest)| {
        let q = seconds / (time + rest);
        let r = if *time < (seconds % (*time + *rest)) {*time} else {seconds % (*time + *rest)} ;
        q * time * speed + speed*  r
    }
)
.collect_vec()
}